sstp_protocol/src/response.rs

```diff
@@ -0,0 +1,397 @@
+    ///
+    /// ```rust
+    /// # use sstp_protocol::response::AdditionalData;
+    /// let data = AdditionalData::from_static("sakura").unwrap();
+    /// matches!(data, AdditionalData::Text(bytes) if bytes == b"sakura");
+    /// ```
+    pub fn from_static(s: &str) -> Result<Self, encode::Error> {
+        Self::from_static_with_charset(s, Charset::ASCII)
+    }
+
+    ///　Convert string to AdditionalData with Charset.
+    ///
+    /// ```rust
+    /// # use sstp_protocol::Charset;
+    /// # use sstp_protocol::response::AdditionalData;
+    /// let data = AdditionalData::from_static_with_charset("さくら", Charset::SHIFT_JIS).unwrap();
+    /// matches!(data, AdditionalData::Text(bytes) if bytes == [130, 179, 130, 173, 130, 231]);
+    /// ```
+    pub fn from_static_with_charset(s: &str, charset: Charset) -> Result<Self, encode::Error> {
+        if s.is_empty() {
+            Ok(Self::Empty)
+        } else {
+            let s = s.replace('\n', "\r\n");
+            let mut bytes = match charset {
+                Charset::ASCII => Encoder::encode_ascii(s.as_str())?,
+                Charset::SHIFT_JIS => Encoder::encode_sjis(s.as_str())?,
+                Charset::ISO2022JP => Encoder::encode_iso_2022_jp(s.as_str())?,
+                Charset::EUC_JP => Encoder::encode_euc_jp(s.as_str())?,
+                Charset::UTF8 => Encoder::encode_utf8(s.as_str())?,
+            };
+            if &bytes[(bytes.len() - 1)..bytes.len()] != b"\r\n" {
+                bytes.extend_from_slice(b"\r\n");
+            }
+            Ok(Self::Text(bytes))
+        }
+    }
+}
+
+impl From<&[u8]> for AdditionalData {
+    fn from(s: &[u8]) -> Self {
+        Self::from(s.to_vec())
+    }
+}
+
+impl From<Vec<u8>> for AdditionalData {
+    fn from(s: Vec<u8>) -> Self {
+        if s.is_empty() {
+            Self::Empty
+        } else {
+            Self::Text(s)
+        }
+    }
+}
+
+impl Default for AdditionalData {
+    fn default() -> Self {
+        Self::Empty
+    }
+}
+
+/// Response is a type that represents an SSTP response.
+///
+/// Response provides a builder to generate types, a parser to generate types from bytes.
+///
+/// ```rust
+/// # use sstp_protocol::{Charset, HeaderName, Method, StatusCode, Version};
+/// # use sstp_protocol::response::Response;
+/// let response = Response::builder()
+///     .version(Version::SSTP_11)
+///     .status_code(StatusCode::OK)
+///     .charset(Charset::UTF8)
+///     .header(
+///         HeaderName::SCRIPT,
+///         "\\h\\s0テストー。\\u\\s[10]テストやな。",
+///     )
+///     .additional("追加データはここ")
+///     .build()
+///     .unwrap();
+/// assert_eq!(response.status_code(), StatusCode::OK);
+/// ```
+#[derive(Debug)]
+pub struct Response {
+    pub(crate) version: Version,
+    pub(crate) status_code: StatusCode,
+    pub(crate) headers: HeaderMap,
+    pub(crate) charset: Charset,
+    pub(crate) additional: AdditionalData,
+}
+
+impl Response {
```

Comment:
Yes, SSTP is a protocol inspired by HTTP, and I think there are places where it can be reused.
However, I do not fully understand the difference between SSTP and HTTP, so I believe it is less troublesome to create a new one than to reuse it.